plugins/workflow/category/src/Extension/Category.php

```diff
@@ -0,0 +1,259 @@
+            });
+        ";
+        $this->getApplication()->getDocument()->addScriptDeclaration($js);
+    }
+
+
+    /**
+     * Method to handle the workflow transition event.
+     *
+     * @param   WorkflowTransitionEvent  $event  The event object
+     *
+     * @return  void
+     *
+     * @since   DEPLOY_VERSION
+     */
+    public function onWorkflowAfterTransition(WorkflowTransitionEvent $event): void
+    {
+        $app = $this->getApplication();
+        $pks = $event->getArgument('pks');
+        $transition = $event->getArgument('transition');
+
+        if (!is_object($transition) || !($transition->options instanceof Registry) || empty($transition->options)) {
+            $app->enqueueMessage('PLG_WORKFLOW_CATEGORY_INVALID_TRANSITION');
+            return;
+        }
+
+        $options = $transition->options;
+        $categoryId = (int) $options->get('category_id');
+
+        if (empty($pks) || !is_array($pks)) {
+            $app->enqueueMessage(Text::_('PLG_WORKFLOW_CATEGORY_NO_PRIMARY_KEY'), 'error');
+            return;
+        }
+
+        $errors = 0;
+
+        foreach ($pks as $pk) {
+            if (!self::processArticle($pk, $categoryId)) {
+                $errors++;
+            }
+        }
+
+        if ($errors > 0) {
+            $app->enqueueMessage(Text::plural('PLG_WORKFLOW_CATEGORY_ERRORS', $errors), 'warning');
+        }
+    }
+
+    /**
+     * Process the article and update its category.
+     *
+     * @param   int                  $pk         The primary key
+     * @param   int                  $categoryId The category ID
+     *
+     * @return  bool
+     *
+     * @since   DEPLOY_VERSION
+     */
+    private function processArticle($pk, $categoryId): bool
+    {
+        $app = $this->getApplication();
+        $result = false;
+
+        try {
+
+            $component = $this->getApplication()->bootComponent('com_content');
+            $modelName = $component->getModelName('com_workflow.article');
+
+            $articleTable = $component->getMVCFactory()->createModel($modelName, $this->getApplication()->getName(), ['ignore_request' => true])
+                ->getTable();
+            if (!$articleTable->load($pk)) {
+                $app->enqueueMessage(Text::sprintf('PLG_WORKFLOW_CATEGORY_ARTICLE_NOT_FOUND', $pk), 'warning');
+                return false;
+            } elseif ($articleTable->catid == $categoryId) {
+                $result = true;
+            } else {
+                $originalData = clone $articleTable;
+                if ($categoryId && $categoryId > 0) {
+                    $articleTable->catid = $categoryId;
+                }
+
+                $articleTable->modified = $originalData->modified;
+                $articleTable->modified_by = $originalData->modified_by;
+
+                if (!$articleTable->store()) {
+                    $app->enqueueMessage(Text::sprintf('PLG_WORKFLOW_CATEGORY_ARTICLE_UPDATE_FAILED', $pk, $categoryId), 'error');
```

Comment:
it's good to be that verbose, but you should not set all this enqueue message, otherwise the user is flooded. If something breaks, throw an exception and catch it somewhere later where you generate the minimum amount of messages (but as many as needed).